### PR TITLE
[RNMobile] Long-press on inserter to show options for "add above" and "add below"

### LINKED ISSUE #1433: Proposal: Long-press on inserter to show options for "add above" and "add below"
Long-pressing on the inserter button (`(+)` on the quick toolbar) should show options to `Add Block Before` and `Add Block After` the currently selected block. If there is no selection these options could be `Add to beginning` and `Add to end`. On iOS, this could be an `ActionSheet` and Android a `BottomSheet`.

Example:

IMG_0645

Note: I think we could expand on this pattern in the future, but I've found myself wanting to do add blocks before the selected one quite often, especially now that we have InnerBlocks.

We might also want to add these options to the `•••` menu on the inline toolbar when we get a chance, similar to web.

### PR DESCRIPTION
**Pull request by ceyhun**
## Description

Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1433

## How has this been tested?

I've tested it using iPhone X Simulator with iOS 13.2 and Pixel 3 Android emulator with API 28.

### Elements
1. Inserter button (+) behavior
(The current behavior on normal press should not change)
1. "ADD BLOCK HERE" indicator behavior
1. Bottom/ActionSheet behavior

### Steps

1. Long press on the inserter button (+)
1. Bottom sheet should be presented 
1. Options should should be different depending on the test case
1. Select an option
1. "ADD BLOCK HERE" indicator should be shown in the correct place depending on the option you picked
1. Menu to pick a block should be shown
1. Select a block
1. Block should be inserted in the correct place

### Cases

1. When no block is selected
1. When title is selected  
    - Bottom sheet options should be:  
    `Add to beginning`  
    `Add to end`
1. When unmodified default block (`Paragraph`) is selected  
    - Unmodified default block (`Paragraph`) should be kept as it is and insertion should happen according to the option that's picked.
1. When a block other than `Paragraph` is selected  
1. When a block in a `Media & Text` block is selected
1. When a block in a `Group` is selected  
**Note**: Appending a block in an empty group currently doesn’t work. There’s a PR to fix that: https://github.com/WordPress/gutenberg/pull/18564. A group block with some inner blocks can be added locally to `initial-html` for testing.
    - Bottom sheet options should be:  
    `Add Block Before`  
    `Add Block After`

### Nothing selected

iOS | Android
--  | --
no-selection-ios | no-selection-android

### Unmodified `Paragraph` selected

iOS | Android
--  | --
unmodified-default-ios | unmodified-default-android

### `Image` selected  

iOS | Android
--  | --
image-selected-ios | image-selected-android

### Inside `Media & Text` 

iOS | Android
--  | --
media-text-block-ios | media-text-block-android

### Inside `Group` 

iOS | Android
--  | --
in-group-block-selected-ios | in-group-block-selected-android

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by ceyhun** [comment]
**Current state of things:** The behaviour of “ADD BLOCK HERE” indicator when replacing/keeping an unmodified default block required a bit more work. I implemented it by adding a new variable to the `insertionPoint` state/reducer/action. Now it doesn’t replace an unmodified default block when inserting before it. Also now the `InserterMenu` is shown on long press and the block you pick is inserted before.
Next step is to show the Bottom/ActionSheet on long press, so you can choose to insert before or after.

**Comment by ceyhun** [comment]
Implemented all the required functionality. Now it also shows the Bottom/ActionSheet on long press and you can choose to insert before or after.

**Comment by SergioEstevao** [comment]
I've tested this on iOS and is working as intended.

The only little thing that I noticed is that when we add the new block to out of the screen the app doesn't scroll to that location. This may have been an already existing bug, but if possible to arrange on this PR will be great.

Code wise is looking solid!

**Comment by iamthomasbishop** [comment]
@ceyhun I've got a few bits of feedback as I'm working through a design review.

* The actions list on the Bottom Sheet on Android should be left-aligned to the 16px keyline
* You can remove the `Cancel` button from the Bottom Sheet on Android
* On the last example (when long-pressing on the Inserter from an empty new block), I noticed something we might want to consider – adding a  `Replace Current Block` option in addition to the before/after options when showing the Action/Bottom Sheet. If we add this, I would add it in the in the middle of the list (`Add to Beginning`, `Replace Current Block`, `Add to End`)

**Comment by ceyhun** [comment]
> The only little thing that I noticed is that when we add the new block to out of the screen the app doesn't scroll to that location. This may have been an already existing bug, but if possible to arrange on this PR will be great.

Sure! I'll check what’s the underlying issue and if it makes sense to include it in this PR.

**Comment by ceyhun** [comment]
@iamthomasbishop thanks for the feedback!

> * The actions list on the Bottom Sheet on Android should be left-aligned to the 16px keyline  

How about also displaying the icons for Android? It will look like this in the end:

 

> * You can remove the Cancel button from the Bottom Sheet on Android  

👍 

> * On the last example (when long-pressing on the Inserter from an empty new block), I noticed something we might want to consider – adding a  `Replace Current Block` option in addition to the before/after options when showing the Action/Bottom Sheet. If we add this, I would add it in the in the middle of the list (`Add to Beginning`, `Replace Current Block`, `Add to End`)

In this case should we show the `Replace Current Block` option only for when an unmodified/empty paragraph block is selected or also for other types of blocks as well when they are not yet modified?
Also I think this option doesn’t exist on web editor menus and if we decide to show icons on Android what can we use for this option?

**Comment by iamthomasbishop** [comment]
> How about also displaying the icons for Android? It will look like this in the end:

Yes! Absolutely 👍 

> In this case should we show the Replace Current Block option only for when an unmodified/empty paragraph block is selected or also for other types of blocks as well when they are not yet modified?

Ah, good question. I was mainly thinking empty Paragraph blocks, but other RichText blocks such as Heading should probably behave the same way. For example, if I add a heading block, leave it empty and come back to select it, it'd seem intuitive to me if the same logic applied to a Heading. 

Looking at the full list of available blocks, we could probably apply the same to other RichText blocks like _Lists_, _Quote_,  _Code_, and _Pre_ blocks, but I think those would be less intuitive than simple blocks like _Paragraph_ and _Heading_. There are some that wouldn't make sense, like media blocks (_Image_, _Video_), _More_, _Page Break,_ _Separator_, _Spacer_, _Media & Text_, etc.

**Comment by ceyhun** [comment]
@iamthomasbishop Other blocks than Paragraph could be a little tricky as we may need to keep an up-to-date list of “replaceable” blocks and add new blocks there as they become available. How about only making paragraph replaceable for now? It would also be consistent with the behavior of normal-press to the inserter (+) because only empty paragraph blocks are replaced on normal-press right now.

For the `Replace Current Block` option's icon I tried with this one from the web:
replace-icon

It looks like this on Android. What do you think?
replace-menu

**Comment by ceyhun** [comment]
@SergioEstevao about scrolling to a new block that’s out of the screen:

At first I looked if it was working with blocks that open the keyboard.

Both iOS and Android use `KeyboardAwareFlatList` component to render the `BlockList`.
But inside `KeyboardAwareFlatList`:
- iOS uses `react-native-keyboard-aware-scroll-view`
- Android [uses `KeyboardAvoidingView` from "react-native"](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/mobile/keyboard-aware-flat-list/index.android.js#L12) but a `behavior` prop is not passed in so [it acts as a normal `View`](https://github.com/jtreanor/react-native/blob/master/Libraries/Components/Keyboard/KeyboardAvoidingView.js#L218)



For both iOS and Android 
- With blocks that open the keyboard:
  - If the new block is at the **top** of the current render window in the list it **doesn’t work**.
  - If the new block is at the **bottom** of the current render window in the list it **works**.

- With blocks that don't open keyboard it doesn’t work at all as you noticed.

So there were 2 cases where it wasn’t working:
1. New block that **does not** open keyboard
2. New block that does open keyboard but is at the **top** of the current render window


Then I looked into [FlatList.scrollToIndex](https://facebook.github.io/react-native/docs/flatlist#scrolltoindex) method to solve this. The documentation says: 

> Cannot scroll to locations outside the render window without specifying the `getItemLayout` prop.

The `getItemLayout` needs every item’s height calculated in advance. It might be tricky to calculate every block’s height beforehand. So I didn’t go with the FlatList solution.

Finally I looked into the `react-native-keyboard-aware-scroll-view` a bit. I realised that it has a `scrollIntoView` method. It takes a ref as a parameter and scrolls it into the view. So I used this method. Also used the same component for Android to be able to call this method. Now this works for both cases and both platforms.

iOS | Android
--  | --
scroll-to-end-ios | scroll-to-end-android

**Comment by hypest** [comment]
Nice to see `react-native-keyboard-aware-scroll-view` working for Android @ceyhun , thanks! I wonder though, perhaps there were some issues we encountered in the past that prevented us from using it for Android. @pinarol , do you perhaps recall why we ended up only using the library for iOS and not for Android?

**Comment by iamthomasbishop** [comment]
>  Other blocks than Paragraph could be a little tricky as we may need to keep an up-to-date list of “replaceable” blocks and add new blocks there as they become available

@ceyhun This is a good point regarding scalability. Let's just start w/ paragraph as this is the most common part.

> For the Replace Current Block option's icon I tried with this one from the web:

I would avoid using the same icon that is already used for re-usable blocks. Let's use the simple `(+)` icon. 

Also, the scrolling behavior you shared above looks great. I know we've had some technical issues with scrolling behavior but improving that aspect of the editor has been a design-wish of mine so hopefully this helps 😄

**Comment by pinarol** [comment]
>I wonder though, perhaps there were some issues we encountered in the past that prevented us from using it for Android. @pinarol , do you perhaps recall why we ended up only using the library for iOS and not for Android?

We simply didn't needed that for Android because Android was by default able to scroll to the focused text field when the keyboard opens. So, `KeyboardAvoidingView` was just enough. Also, it was written in the README that Android support of `react-native-keyboard-aware-scroll-view` wasn't good. From what I have seen this PR uses it with `enableOnAndroid={ false }` so it will continue using Android's default autoscroll capabilities. The thing that worries me more is dropping `KeyboardAvoidingView` and not putting something that replicates its functionality. I am a bit confused how it can be possible because then this means `KeyboardAvoidingView` was not necessary at all 🤔OR we are missing something.

What I can suggest is trying to avoid dropping `KeyboardAvoidingView` just for being able to call [scrollIntoView](https://github.com/wordpress-mobile/react-native-keyboard-aware-scroll-view/blob/6bc6f97488cf57c2277612c361f82bd960ff25b6/lib/KeyboardAwareHOC.js#L337). We can consider adding `scrollIntoView` function in our code base and call it by passing scrollView's ref to it. This could minimize possible future regressions. It seems like a pretty standalone function to me that doesn't use a lot from that lib.

**Comment by ceyhun** [comment]
> I would avoid using the same icon that is already used for re-usable blocks. Let's use the simple `(+)` icon.

@iamthomasbishop You mean this one?  It's used for the inserter button as well.

**Comment by ceyhun** [comment]
> I am a bit confused how it can be possible because then this means `KeyboardAvoidingView` was not necessary at all 🤔OR we are missing something.

@pinarol on Android a `behavior` prop is not passed into the `KeyboardAvoidingView` 
 https://github.com/WordPress/gutenberg/blob/master/packages/components/src/mobile/keyboard-aware-flat-list/index.android.js#L12

I think it acts as a normal `View` when a `behavior` prop is not passed in. But I'm not 100% sure 😄 
https://github.com/jtreanor/react-native/blob/master/Libraries/Components/Keyboard/KeyboardAvoidingView.js#L218

> What I can suggest is trying to avoid dropping `KeyboardAvoidingView` just for being able to call [scrollIntoView](https://github.com/wordpress-mobile/react-native-keyboard-aware-scroll-view/blob/6bc6f97488cf57c2277612c361f82bd960ff25b6/lib/KeyboardAwareHOC.js#L337). We can consider adding `scrollIntoView` function in our code base and call it by passing scrollView's ref to it. This could minimize possible future regressions. It seems like a pretty standalone function to me that doesn't use a lot from that lib.

On Android we can still keep the `KeyboardAvoidingView` to avoid causing regressions. But in this case the only scrollable element inside `KeyboardAwareFlatList` remains the `FlatList`. I can try to extract `scrollIntoView` function into `KeyboardAwareFlatList` and pass the `FlatList`'s ref to it. It might still work since `FlatList` also inherits `ScrollView` props.

**Comment by pinarol** [comment]
> I think it acts as a normal View when a behavior prop is not passed in. But I'm not 100% sure 😄
https://github.com/jtreanor/react-native/blob/master/Libraries/Components/Keyboard/KeyboardAvoidingView.js#L218

In that case I wouldn't worry. One thing to make sure though, we are not using react-native's `KeyboardAvoidingView` directly from this file, we have our own wrapper on top of it, so let's check if we don't use `behavior` there either.

> It might still work since FlatList also inherits ScrollView props.

Makes sense but I am also OK with the current state if we make sure KeyboardAvoidingView acts no different than a normal View in this case.

**Comment by iamthomasbishop** [comment]
> You mean this one?
Yep that one. That's okay, it makes sense to use the same icon as it essentially performs the default inserter action.

**Comment by SergioEstevao** [comment]
> @SergioEstevao about scrolling to a new block that’s out of the screen:
> 
> At first I looked if it was working with blocks that open the keyboard.
> 
> Both iOS and Android use `KeyboardAwareFlatList` component to render the `BlockList`.
> But inside `KeyboardAwareFlatList`:
> 
> * iOS uses `react-native-keyboard-aware-scroll-view`
> * Android [uses `KeyboardAvoidingView` from "react-native"](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/mobile/keyboard-aware-flat-list/index.android.js#L12) but a `behavior` prop is not passed in so [it acts as a normal `View`](https://github.com/jtreanor/react-native/blob/master/Libraries/Components/Keyboard/KeyboardAvoidingView.js#L218)
> 
> For both iOS and Android
> 
> * With blocks that open the keyboard:
>   
>   * If the new block is at the **top** of the current render window in the list it **doesn’t work**.
>   * If the new block is at the **bottom** of the current render window in the list it **works**.
> * With blocks that don't open keyboard it doesn’t work at all as you noticed.
> 
> So there were 2 cases where it wasn’t working:
> 
> 1. New block that **does not** open keyboard
> 2. New block that does open keyboard but is at the **top** of the current render window
> 
> Then I looked into [FlatList.scrollToIndex](https://facebook.github.io/react-native/docs/flatlist#scrolltoindex) method to solve this. The documentation says:
> 
> > Cannot scroll to locations outside the render window without specifying the `getItemLayout` prop.
> 
> The `getItemLayout` needs every item’s height calculated in advance. It might be tricky to calculate every block’s height beforehand. So I didn’t go with the FlatList solution.
> 
> Finally I looked into the `react-native-keyboard-aware-scroll-view` a bit. I realised that it has a `scrollIntoView` method. It takes a ref as a parameter and scrolls it into the view. So I used this method. Also used the same component for Android to be able to call this method. Now this works for both cases and both platforms.
> 
> iOS	Android
> scroll-to-end-ios	scroll-to-end-android

@ceyhun tested this on iOS and now it's working perfectly.

File: packages/block-editor/src/components/inserter/index.native.js [hunk 1]
```diff
@@ -59,7 +134,42 @@ class Inserter extends Component {
 			getStylesFromColorScheme,
 		} = this.props;
 		const style = getStylesFromColorScheme( styles.addBlockButton, styles.addBlockButtonDark );
-		return renderToggle( { onToggle, isOpen, disabled, style } );
+
+		const onPress = () => {
+			this.setState( { insertionType: 'default' }, () => {
```
Comment (Tug) [inline packages/block-editor/src/components/inserter/index.native.js]: `this.setState( { insertionType: 'default' }, onToggle );` is a bit shorter

File: packages/block-editor/src/components/inserter/index.native.js [hunk 2]
```diff
@@ -59,7 +134,42 @@ class Inserter extends Component {
...
+		};
+
+		const onLongPress = () => {
+			if ( this.picker ) {
+				this.picker.presentPicker();
+			}
+		};
+
+		const onPickerSelect = ( value ) => {
+			this.setState( { insertionType: value }, () => {
```
Comment (Tug) [inline packages/block-editor/src/components/inserter/index.native.js]: same here

File: packages/block-editor/src/components/inserter/index.native.js [hunk 3]
```diff
@@ -101,23 +216,71 @@ class Inserter extends Component {
...
 		} = select( 'core/block-editor' );
 
-		let destinationRootClientId = rootClientId;
-		if ( ! destinationRootClientId && ! clientId && ! isAppender ) {
-			const end = getBlockSelectionEnd();
-			if ( end ) {
-				destinationRootClientId = getBlockRootClientId( end ) || undefined;
+		const end = getBlockSelectionEnd();
+		const isAnyBlockSelected = ( ! isAppender && end );
+		const destinationRootClientId = isAnyBlockSelected ?
```
Comment (Tug) [inline packages/block-editor/src/components/inserter/index.native.js]: Looks like a fix! Was it not inserting at the right place before that?

File: packages/components/src/button/index.native.js [hunk 4]
```diff
@@ -105,6 +106,7 @@ export function Button( props ) {
 			accessibilityRole={ 'button' }
 			accessibilityHint={ hint }
 			onPress={ onClick }
+			onLongPress={ onLongPress }
```
Comment (Tug) [inline packages/components/src/button/index.native.js]: Nice ❤️

**Review by Tug — COMMENTED** [review]
Nice fix for the autoscroll to block, I tested it and it works well 👍 
Would have been best to do it in another PR imo as I feel this PR is quite complex already. At least it looks complex to me because of the changes of state we introduce, I find it a bit hard to follow across the different components. Anyway it's ok to have it here since Sergio asked for it ;)

I'd like us to think of an alternative approach that does not involve modifying the web store. Let's try to keep the insertionPoint reducer "UI only" if possible.

File: packages/block-editor/src/store/reducer.js [hunk 5]
```diff
@@ -1121,8 +1121,8 @@ export function blocksMode( state = {}, action ) {
 export function insertionPoint( state = null, action ) {
 	switch ( action.type ) {
 		case 'SHOW_INSERTION_POINT':
-			const { rootClientId, index } = action;
-			return { rootClientId, index };
+			const { rootClientId, index, canReplaceBlock } = action;
```
Comment (jorgefilipecosta) [inline packages/block-editor/src/store/reducer.js]: Hi @ceyhun, it seems canReplaceBlock is not used as part of the state logic in the reducer, it just passed in action so it can be retrieved from the state later.
I wonder if instead of storing canReplaceBlock in the state it would be possible to call something equivalent to getShouldReplaceBlock() when we need this information?

Comment (jorgefilipecosta) [reply packages/block-editor/src/store/reducer.js]: This goes in a similar direction to what @Tug was recommending.

File: packages/block-editor/src/components/inserter/index.native.js [hunk 6]
```diff
@@ -28,11 +58,56 @@ class Inserter extends Component {
...
+		if ( insertionType === 'before' || insertionType === 'replace' ) {
+			return insertionIndexBefore;
+		}
+		if ( insertionType === 'after' ) {
+			return insertionIndexAfter;
+		}
+		return insertionIndexDefault;
+	}
+
+	getShouldReplaceBlock() {
```
[the remark this answers is not in the document]
Comment (ceyhun) [reply packages/block-editor/src/components/inserter/index.native.js]: 👍

File: packages/block-editor/src/components/inserter/index.native.js (hunk 1 above)
Comment (ceyhun) [reply packages/block-editor/src/components/inserter/index.native.js]: 👍

File: packages/block-editor/src/components/inserter/index.native.js (hunk 2 above)
Comment (ceyhun) [reply packages/block-editor/src/components/inserter/index.native.js]: 👍

File: packages/components/src/button/index.native.js (hunk 4 above)
Comment (ceyhun) [reply packages/components/src/button/index.native.js]: Thanks 🙂

File: packages/block-editor/src/components/inserter/index.native.js (hunk 3 above)
Comment (ceyhun) [reply packages/block-editor/src/components/inserter/index.native.js]: Previously `withSelect` implementation of `Inserter` was returning `items` and it was not used at all. The changes are actually not related with each other.

File: docs/designers-developers/developers/data/data-core-block-editor.md [hunk 7]
```diff
@@ -1172,6 +1172,7 @@ _Parameters_
 
 -   _rootClientId_ `?string`: Optional root client ID of block list on which to insert.
 -   _index_ `?number`: Index at which block should be inserted.
+-   _canReplaceBlock_ `?boolean`: can the insertion point replace the block at index.
```
[the remark this answers is not in the document]
Comment (ceyhun) [reply docs/designers-developers/developers/data/data-core-block-editor.md]: It is already derived from `Inserter`’s local state via [getShouldReplaceBlock](https://github.com/WordPress/gutenberg/pull/18791/files#diff-8aa4726165b0e0cdb260abe61e3329ceR97) function. But we need to pass this state to the `BlockList` so it [hides the blocks](https://github.com/WordPress/gutenberg/pull/18791/files#diff-260b0f5b413f5ac735a6935b2f8acc34R170) where the `BlockInsertionPoint` will replace. `Inserter` and `BlockList` don't share a very close parent so instead of lifting the state all the way up to their common parent and drilling down, I chose to store this state it in the store :) It might be used in web for a similar behavior with "Add Block Before/After" for where a block in the list could be hidden when it will be replaced and `BlockInsertionPoint` is shown in it's place.

But I agree this state is now not about only showing/hiding `BlockInsertionPoint` but it's even about showing/hiding a block in the list.  

But maybe there is another way without changing the web store. Instead of hiding a block we can try to remove the block completely from `BlockList` when `BlockInsertionPoint` is shown and the block is going to be replaced. If the block insertion is cancelled (via closing the menu) we can re-insert the block again. It should be easy to re-insert the block again because the replace only happens for unmodified default blocks (empty paragraph) right now. If a new block is picked from the menu instead of cancelling we can just call [insert](https://github.com/WordPress/gutenberg/pull/18791/files#diff-e7138e42f97fc785033583bf87f62519R166) instead of [replace](https://github.com/WordPress/gutenberg/pull/18791/files#diff-e7138e42f97fc785033583bf87f62519R164) to add the new block. This way we might even remove [show/hide block](https://github.com/WordPress/gutenberg/pull/18791/files#diff-260b0f5b413f5ac735a6935b2f8acc34R262) logic completely from `BlockList`.  

What do you think?

File: packages/block-editor/src/store/reducer.js (hunk 5 above)
Comment (ceyhun) [reply packages/block-editor/src/store/reducer.js]: Hi @jorgefilipecosta, it is a state deriving from user picking “Add Before/After“ or “Replace” and it is kept locally in the `Inserter`. Later on it’s used in `BlockList` to hide the block that could be replaced. I’m thinking about another approach. Please see my other comment [here](https://github.com/WordPress/gutenberg/pull/18791#discussion_r358707205).

**Comment by ceyhun** [comment]
> Would have been best to do it in another PR imo as I feel this PR is quite complex already. At least it looks complex to me because of the changes of state we introduce, I find it a bit hard to follow across the different components. Anyway it's ok to have it here since Sergio asked for it ;)

android-scrollview-bug


I just found this bug with group blocks on Android now after merging `develop` into `guteberg-mobile` and rebasing my changes here onto `guteberg-mobile/develop`’s gutenberg ref. Sometimes the groups are not being rendered correctly. It seems to be caused by using `react-native-keyboard-aware-scroll-view` on Android because without that it just works fine. 

We might need a more complex solution that doesn’t involve `react-native-keyboard-aware-scroll-view` for Android. @SergioEstevao For this I think it makes more sense to do it in another PR too.

**Comment by pinarol** [comment]
> Sometimes the groups are not being rendered correctly. It seems to be caused by using react-native-keyboard-aware-scroll-view on Android because without that it just works fine.

>We might need a more complex solution that doesn’t involve react-native-keyboard-aware-scroll-view for Android. @SergioEstevao For this I think it makes more sense to do it in another PR too.

[InnerBlock enhancements](https://github.com/WordPress/gutenberg/pull/17547) are merged recently and it was a big change, it could have effected this PR as well. I didn't quite understand the buggy behavior from that screenshot but I'd avoid merging this if it is introducing some problems on Group block. I'd also suggest checking Media-Text because it uses InnerBlocks like Group, and it is already on prod. We want to ship Column and Group very soon so we wouldn't want to introduce blockers for them.

> Sometimes the groups are not being rendered correctly

Could you elaborate on this a bit? What kind of problem are you observing?

Also, there was this one suggestion that you made @ceyhun, would it help solving the issue?

> On Android we can still keep the KeyboardAvoidingView to avoid causing regressions. But in this case the only scrollable element inside KeyboardAwareFlatList remains the FlatList. I can try to extract scrollIntoView function into KeyboardAwareFlatList and pass the FlatList's ref to it. It might still work since FlatList also inherits ScrollView props.

react-native-keyboard-aware-scroll-view can be buggy on Android, the developers of the original library made it clear in the README that Android support wasn't perfect.

**Comment by ceyhun** [comment]
> Could you elaborate on this a bit? What kind of problem are you observing?

The Paragraphs in Group block doesn’t render initially unless you add another block at the bottom (Image in this case). When the Image block is removed they stay rendered.

> I'd also suggest checking Media-Text because it uses InnerBlocks like Group, and it is already on prod.

Similar thing happens with Media & Text as well. But in a different way: paragraph blocks only render when the Media & Text block is selected and disappear when you you select and unselect another block.

> Also, there was this one suggestion that you made @ceyhun, would it help solving the issue?

It might solve the issue but I’m afraid it could make this PR a bit complicated and could introduce a large changeset in a single PR. Also I can potentially introduce some more changes related to `Inserter` and `BlockList` that I mentioned [here](https://github.com/WordPress/gutenberg/pull/18791#discussion_r358707205) that are not related with ScrollView. Without the “_scroll to newly added block_” feature everything works as expected. Maybe it's better to introduce “_scroll to newly added block_” feature in another PR.

**Comment by pinarol** [comment]
> The Paragraphs in Group block doesn’t render initially unless you add another block at the bottom (Image in this case). When the Image block is removed they stay rendered.

> Similar thing happens with Media & Text as well. But in a different way: paragraph blocks only render when the Media & Text block is selected and disappear when you you select and unselect another block.

Hmm these issues are real blockers we need to avoid.

> Without the “scroll to newly added block” feature everything works as expected. Maybe it's better to introduce “scroll to newly added block” feature in another PR.

I like this idea. 👍Although let's hear from @hypest also.

**Comment by hypest** [comment]
> >    Without the “scroll to newly added block” feature everything works as expected. Maybe it's better to introduce “scroll to newly added block” feature in another PR.
>
> I like this idea. 👍Although let's hear from @hypest also.

I think that scrolling to the block is outside the scope of the ticket/task this PR is after, and since it seems to introduce side effects, I'd say let's go with removing it and tackling it in a separate PR @ceyhun .

File: docs/designers-developers/developers/data/data-core-block-editor.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tug) [reply docs/designers-developers/developers/data/data-core-block-editor.md]: > What do you think?

I think it sounds quite hacky, but that's actually a clever way to do it without compromising on the functionalities, I like it 👍

I say it's hacky because this approach won't work if `isSelectedBlockReplaceable` becomes more complex than just `isSelectedUnmodifiedDefaultBlock`. Maybe we will then be able to "store" the type of block that was removed in the Inserter state, but then it feels even more hacky 🤔

I think it's fine for now to have more custom code in the native Inserter since our version does more than the web. Adding new attributes to the store that are not presently needed is hardly justifiable at this point in the project.

**Comment by ceyhun** [comment]
> I think that scrolling to the block is outside the scope of the ticket/task this PR is after, and since it seems to introduce side effects, I'd say let's go with removing it and tackling it in a separate PR @ceyhun .

👍 Reverted the scroll to block functionality.

File: docs/designers-developers/developers/data/data-core-block-editor.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ceyhun) [reply docs/designers-developers/developers/data/data-core-block-editor.md]: I just pushed the mentioned changes (and also reverted scrollview related changes), feel free to check it out again :)

File: packages/block-editor/src/components/inserter/index.native.js [hunk 8]
```diff
@@ -2,19 +2,49 @@
...
+import { isUnmodifiedDefaultBlock } from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import styles from './style.scss';
 import InserterMenu from './menu';
 
-const defaultRenderToggle = ( { onToggle, disabled, style } ) => (
+const addBeforeOption = {
```
Comment (Tug) [inline packages/block-editor/src/components/inserter/index.native.js]: Not sure it's worth it having them module scoped, they could be defined inside `getInsertionOptions`

**Review by Tug — CHANGES_REQUESTED** [review]
Tested on Android and it works well 👍 
Code looks solid as well, I only had a few minor notes and questions

File: packages/block-editor/src/components/block-list/index.native.js [hunk 9]
```diff
@@ -120,27 +120,19 @@ export class BlockList extends Component {
...
-	}
-
-	renderItem( { item: clientId, index } ) {
-		const { shouldShowBlockAtIndex, shouldShowInsertionPointBefore, shouldShowInsertionPointAfter } = this.props;
+	renderItem( { item: clientId } ) {
+		const { shouldShowInsertionPointBefore, shouldShowInsertionPointAfter } = this.props;
 		return (
 			<ReadableContentView>
 				{ shouldShowInsertionPointBefore( clientId ) && this.renderAddBlockSeparator() }
-				{ shouldShowBlockAtIndex( index ) && (
```
[the remark this answers is not in the document]
Comment (ceyhun) [reply packages/block-editor/src/components/block-list/index.native.js]: 🎉

File: packages/block-editor/src/components/inserter/menu.native.js [hunk 10]
```diff
@@ -140,81 +139,59 @@ export default compose(
...
+					const {
+						getBlockOrder,
+						getBlockCount,
+					} = select( 'core/block-editor' );
+
+					const count = getBlockCount();
+					if ( count === 1 ) {
+						clearSelectedBlock();
+						resetBlocks( [] );
+					}	else {
```
[the remark this answers is not in the document]
Comment (ceyhun) [reply packages/block-editor/src/components/inserter/menu.native.js]: 👍

File: packages/block-editor/src/components/inserter/menu.native.js [hunk 11]
```diff
@@ -140,81 +139,59 @@ export default compose(
...
-				const index = getInsertionIndex();
-				showInsertionPoint( ownProps.destinationRootClientId, index );
+				if ( ownProps.shouldReplaceBlock ) {
+					const {
+						getBlockOrder,
+						getBlockCount,
+					} = select( 'core/block-editor' );
+
+					const count = getBlockCount();
+					if ( count === 1 ) {
```
[the remark this answers is not in the document]
Comment (ceyhun) [reply packages/block-editor/src/components/inserter/menu.native.js]: Removing the last block is not possible with `removeBlock` action. It always inserts a default block if the last of the blocks have been removed.
https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/actions.js#L586
I'll add a comment about this here.

File: packages/block-editor/src/components/inserter/index.native.js (hunk 8 above)
Comment (ceyhun) [reply packages/block-editor/src/components/inserter/index.native.js]: 👍

**Review by Tug — APPROVED** [review]
LGTM 👍

File: packages/block-editor/src/components/inserter/menu.native.js [hunk 12]
```diff
@@ -24,18 +21,19 @@ export class InserterMenu extends Component {
...
+		this.onClose = this.onClose.bind( this );
 		this.onLayout = this.onLayout.bind( this );
 		this.state = {
 			numberOfColumns: this.calculateNumberOfColumns(),
 		};
 	}
 
 	componentDidMount() {
-		this.onOpen();
+		this.props.showInsertionPoint();
```
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/inserter/menu.native.js]: In the future refactoring, these components to hooks may increase reusability between mobile and web.
E.g: We can have specific behaviors as hooks like useShowInserPointHook ( hook that trigger the show insert point on the mount and removes it on unmount ) and then the difference between web mobile could simply be the hooks (behaviors) that each one uses.

File: packages/components/src/button/index.native.js [hunk 13]
```diff
@@ -106,6 +107,7 @@ export function Button( props ) {
...
```
Comment (jorgefilipecosta) [inline packages/components/src/button/index.native.js]: Minor non-blocker: The web version of the button components passes down all the props not explicitly used: 		https://github.com/ceyhun/gutenberg/blob/f49806990a8d9219d04c7dbb9cfefb2560f6823d/packages/components/src/button/index.js#L56

Would it make sense to do something equivalent on mobile and simply pass down props to TouchableOpacity without explicitly referring onLongPress?

**Review by jorgefilipecosta — APPROVED** [review]
LGTM 👍 Thank you for iterating on this PR and addressing the concern related to the usage of the global state, the current implementation seems good.

**Review by SergioEstevao — APPROVED** [review]
Working great!
Thanks @ceyhun for tackling all the feedback and back and forward!

File: packages/block-editor/src/components/inserter/menu.native.js (hunk 12 above)
Comment (ceyhun) [reply packages/block-editor/src/components/inserter/menu.native.js]: Hooks ftw 🎉

File: packages/components/src/button/index.native.js (hunk 13 above)
Comment (ceyhun) [reply packages/components/src/button/index.native.js]: Maybe there could be some props passed in that get picked up upper in the tree that we may or may not want to pass to TouchableOpacity directly. 
In that case it could be better to be explicit about it.
